custom/aaa/models.py

```diff
@@ -406,6 +406,594 @@ def agg_from_household_case_ucr(cls, domain, window_start, window_end):
         ), {'domain': domain, 'window_start': window_start, 'window_end': window_end}
 
 
+class AggAwc(models.Model):
+    domain = models.TextField()
+
+    state_id = models.TextField()
+    district_id = models.TextField()
+    block_id = models.TextField()
+    supervisor_id = models.TextField()
+    awc_id = models.TextField()
+
+    month = models.DateField()
+
+    registered_eligible_couples = models.PositiveIntegerField()
+    registered_pregnancies = models.PositiveIntegerField()
+    registered_children = models.PositiveIntegerField()
+
+    eligible_couples_using_fp_method = models.PositiveIntegerField()
+    high_risk_pregnancies = models.PositiveIntegerField()
+    institutional_deliveries = models.PositiveIntegerField()
+    total_deliveries = models.PositiveIntegerField()
+
+    @classmethod
+    def agg_from_woman_table(cls, domain, window_start, window_end):
+        base_tablename = Woman._meta.db_table
+
+        return """
+        INSERT INTO "{agg_tablename}" AS agg (
+            domain, state_id, district_id, block_id, supervisor_id, awc_id, month,
+            registered_eligible_couples,
+            registered_pregnancies,
+            eligible_couples_using_fp_method
+        ) (
+            SELECT
+                %(domain)s,
+                state_id, district_id,
+                block_id, supervisor_id, awc_id,
+                %(window_start)s AS month,
+                COUNT(*) FILTER (WHERE
+                    (NOT %(window_start)s <@ any(pregnant_ranges))
+                    AND migration_status IS DISTINCT FROM 'migrated'
+                    AND marital_status = 'married'
+                ) as registered_eligible_couples,
+                COUNT(*) FILTER (WHERE
+                    %(window_start)s <@ any(pregnant_ranges)
+                    AND migration_status IS DISTINCT FROM 'migrated'
+                ) as registered_pregnancies,
+                COUNT(*) FILTER (WHERE
+                    %(window_start)s <@ any(fp_current_method_ranges)
+                    AND migration_status IS DISTINCT FROM 'migrated'
+                ) as eligible_couples_using_fp_method
+            FROM "{woman_tablename}" woman
+            GROUP BY state_id, district_id, block_id, supervisor_id, awc_id
+        )
+        ON CONFLICT (state_id, district_id, block_id, supervisor_id, awc_id, month) DO UPDATE SET
+           registered_eligible_couples = EXCLUDED.registered_eligible_couples,
+           registered_pregnancies = EXCLUDED.registered_pregnancies,
+           eligible_couples_using_fp_method = EXCLUDED.eligible_couples_using_fp_method
```

Comment:
Are these cumulative counts? Should they be adding to the existing value? Something like

```sql
registered_pregnancies = "{agg_tablename}".registered_pregnancies + EXCLUDED.registered_pregnancies,
```

If yes, it looks like this would also be applicable in other places in this file.

Reply:
These particular columns are going to need to be distinct counts of the cases that apply to the filters. They are aggregating data over the full month at the moment: https://github.com/dimagi/commcare-hq/blob/fa7179f7a20cb723e49ca937637e6b47ce04e611/custom/aaa/tasks.py#L57-L58

I need to read up a bit more on hyperloglog as from that earlier citus article, its possible that it would make it more feasible to do smaller rollups on distinct values